<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Functional GWAS via Quantile Regression and RIF • fungwas</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Functional GWAS via Quantile Regression and RIF">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-primary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">fungwas</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.0.1</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/fungwas-overview.html">Functional GWAS via Quantile Regression and RIF</a></li>
    <li><a class="dropdown-item" href="../articles/weight-building.html">Your custom functional GWAS with `make_weights_generic`</a></li>
  </ul>
</li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Functional GWAS via Quantile Regression and RIF</h1>
            
      

      <div class="d-none name"><code>fungwas-overview.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="motivation">Motivation<a class="anchor" aria-label="anchor" href="#motivation"></a>
</h2>
<p>Most GWAS focus on <strong>mean effects</strong>: does a SNP increase
or decrease the <em>average</em> level of a trait?</p>
<p>But genetic effects can be more subtle:</p>
<ul>
<li>Some SNPs may increase <strong>variability</strong> (vQTLs).</li>
<li>Others may shift the balance between <strong>subtypes or mixture
components</strong>.</li>
<li>Some may affect <strong>both mean and variance</strong>
together.</li>
</ul>
<p>The <code>fungwas</code> package provides a <strong>fast, flexible
framework</strong> to test SNP effects on <em>distributional
parameters</em>, not just means.</p>
<hr>
</div>
<div class="section level2">
<h2 id="core-idea">Core idea<a class="anchor" aria-label="anchor" href="#core-idea"></a>
</h2>
<ol style="list-style-type: decimal">
<li>
<p>Use <strong>Recentered Influence Functions (RIFs)</strong> to
link SNPs to changes in trait quantiles.</p>
<ul>
<li>This gives you “SNP slopes” across the distribution.</li>
<li>It is fast (closed-form OLS, no iterative quantile regression).</li>
</ul>
</li>
<li>
<p>Combine quantile slopes using a <strong>weight matrix W</strong>
to map them into any parameter of interest:</p>
<ul>
<li>mean, variance, mixture membership, etc.</li>
</ul>
</li>
<li><p>Estimate SNP effects and standard errors on those
parameters.</p></li>
</ol>
<p>This two-step process is the backbone of <code>fungwas</code>.</p>
<hr>
</div>
<div class="section level2">
<h2 id="example-1-standard-mean-effect">Example 1: Standard mean effect<a class="anchor" aria-label="anchor" href="#example-1-standard-mean-effect"></a>
</h2>
<p>Let’s simulate a simple trait with SNP effects on the mean.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">N</span> <span class="op">&lt;-</span> <span class="fl">5000</span></span>
<span><span class="va">P</span> <span class="op">&lt;-</span> <span class="fl">50</span></span>
<span><span class="va">maf</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/Uniform.html" class="external-link">runif</a></span><span class="op">(</span><span class="va">P</span>, <span class="fl">0.05</span>, <span class="fl">0.5</span><span class="op">)</span></span>
<span><span class="va">G</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/Binomial.html" class="external-link">rbinom</a></span><span class="op">(</span><span class="va">N</span> <span class="op">*</span> <span class="va">P</span>, <span class="fl">2</span>, <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="va">maf</span>, each <span class="op">=</span> <span class="va">N</span><span class="op">)</span><span class="op">)</span>, <span class="va">N</span>, <span class="va">P</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">G</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"SNP"</span>, <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq_len</a></span><span class="op">(</span><span class="va">P</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Simulate a mean effect</span></span>
<span><span class="va">beta_true</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/Normal.html" class="external-link">rnorm</a></span><span class="op">(</span><span class="va">P</span><span class="op">)</span> <span class="op">/</span> <span class="fl">40</span></span>
<span><span class="va">Y</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/Normal.html" class="external-link">rnorm</a></span><span class="op">(</span><span class="va">N</span>, mean <span class="op">=</span> <span class="fl">2</span> <span class="op">+</span> <span class="va">G</span> <span class="op"><a href="https://rdrr.io/r/base/matmult.html" class="external-link">%*%</a></span> <span class="va">beta_true</span>, sd <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span></span>
<span><span class="va">taus</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span><span class="fl">0.1</span>, <span class="fl">0.9</span>, <span class="fl">0.05</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Stage 1: quantile GWAS</span></span>
<span><span class="va">stage1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/quantile_gwas.html">quantile_gwas</a></span><span class="op">(</span><span class="va">Y</span>, <span class="va">G</span>, taus <span class="op">=</span> <span class="va">taus</span><span class="op">)</span></span>
<span><span class="co">#&gt; Building RIF matrix...</span></span>
<span><span class="co">#&gt; Computing per-SNP tau-slopes...</span></span>
<span></span>
<span><span class="co"># Stage 2: mean/variance mapping (vQTL weights)</span></span>
<span><span class="va">W</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/make_weight_vqtl.html">make_weight_vqtl</a></span><span class="op">(</span><span class="va">taus</span>, <span class="va">stage1</span><span class="op">$</span><span class="va">q_tau</span>, mu <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/mean.html" class="external-link">mean</a></span><span class="op">(</span><span class="va">Y</span><span class="op">)</span>, sd <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/sd.html" class="external-link">sd</a></span><span class="op">(</span><span class="va">Y</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">fit</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/param_gwas.html">param_gwas</a></span><span class="op">(</span><span class="va">stage1</span>, transform <span class="op">=</span> <span class="st">"custom_W"</span>, transform_args <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>W <span class="op">=</span> <span class="va">W</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="va">fit</span><span class="op">$</span><span class="va">params</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt;          beta_mu  beta_sigma2</span></span>
<span><span class="co">#&gt; SNP1 -0.03798818 -0.026820633</span></span>
<span><span class="co">#&gt; SNP2 -0.01832948  0.018785274</span></span>
<span><span class="co">#&gt; SNP3  0.02361505 -0.045563528</span></span>
<span><span class="co">#&gt; SNP4  0.04302755  0.001305224</span></span>
<span><span class="co">#&gt; SNP5  0.05758468 -0.048737974</span></span>
<span><span class="co">#&gt; SNP6  0.01513990  0.199157417</span></span></code></pre></div>
<p>Here, the column <code>beta_mu</code> corresponds to the SNP effect
on the mean.</p>
<hr>
</div>
<div class="section level2">
<h2 id="example-2-variance-effects-vqtls">Example 2: Variance effects (vQTLs)<a class="anchor" aria-label="anchor" href="#example-2-variance-effects-vqtls"></a>
</h2>
<p>Suppose SNPs affect not just the mean but also the
<strong>spread</strong> of the phenotype.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">beta_mu_true</span>     <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/Normal.html" class="external-link">rnorm</a></span><span class="op">(</span><span class="va">P</span><span class="op">)</span> <span class="op">/</span> <span class="fl">40</span></span>
<span><span class="va">beta_sigma2_true</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/Normal.html" class="external-link">rnorm</a></span><span class="op">(</span><span class="va">P</span><span class="op">)</span> <span class="op">/</span> <span class="fl">40</span></span>
<span></span>
<span><span class="va">mu</span> <span class="op">&lt;-</span> <span class="fl">2</span> <span class="op">+</span> <span class="va">G</span> <span class="op"><a href="https://rdrr.io/r/base/matmult.html" class="external-link">%*%</a></span> <span class="va">beta_mu_true</span></span>
<span><span class="va">sigma2</span> <span class="op">&lt;-</span> <span class="fl">1</span> <span class="op">+</span> <span class="va">G</span> <span class="op"><a href="https://rdrr.io/r/base/matmult.html" class="external-link">%*%</a></span> <span class="va">beta_sigma2_true</span></span>
<span><span class="va">sigma2</span><span class="op">[</span><span class="va">sigma2</span> <span class="op">&lt;=</span> <span class="fl">0</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fl">0.1</span></span>
<span></span>
<span><span class="va">Y</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/Normal.html" class="external-link">rnorm</a></span><span class="op">(</span><span class="va">N</span>, mean <span class="op">=</span> <span class="va">mu</span>, sd <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">sqrt</a></span><span class="op">(</span><span class="va">sigma2</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">taus</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span><span class="fl">0.05</span>, <span class="fl">0.95</span>, <span class="fl">0.05</span><span class="op">)</span></span>
<span><span class="va">stage1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/quantile_gwas.html">quantile_gwas</a></span><span class="op">(</span><span class="va">Y</span>, <span class="va">G</span>, taus <span class="op">=</span> <span class="va">taus</span><span class="op">)</span></span>
<span><span class="co">#&gt; Building RIF matrix...</span></span>
<span><span class="co">#&gt; Computing per-SNP tau-slopes...</span></span>
<span></span>
<span><span class="va">W_var</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/make_weight_vqtl.html">make_weight_vqtl</a></span><span class="op">(</span><span class="va">taus</span>, <span class="va">stage1</span><span class="op">$</span><span class="va">q_tau</span>, mu <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/mean.html" class="external-link">mean</a></span><span class="op">(</span><span class="va">Y</span><span class="op">)</span>, sd <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/sd.html" class="external-link">sd</a></span><span class="op">(</span><span class="va">Y</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">fit_var</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/param_gwas.html">param_gwas</a></span><span class="op">(</span><span class="va">stage1</span>, transform <span class="op">=</span> <span class="st">"custom_W"</span>, transform_args <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>W <span class="op">=</span> <span class="va">W_var</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="va">fit_var</span><span class="op">$</span><span class="va">params</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt;          beta_mu beta_sigma2</span></span>
<span><span class="co">#&gt; SNP1 -0.01474599 -0.10471233</span></span>
<span><span class="co">#&gt; SNP2  0.03817296  0.02671172</span></span>
<span><span class="co">#&gt; SNP3 -0.06506831 -0.05781526</span></span>
<span><span class="co">#&gt; SNP4  0.01184463  0.03804201</span></span>
<span><span class="co">#&gt; SNP5 -0.01047337 -0.04669658</span></span>
<span><span class="co">#&gt; SNP6  0.01771871 -0.08678875</span></span></code></pre></div>
<p>Now we obtain SNP effects on both the <strong>mean</strong> and the
<strong>variance</strong>.</p>
<hr>
</div>
<div class="section level2">
<h2 id="example-3-mixture-gwas">Example 3: Mixture GWAS<a class="anchor" aria-label="anchor" href="#example-3-mixture-gwas"></a>
</h2>
<p>Many complex traits are mixtures of underlying subtypes. For example,
“cases” might consist of two symptom clusters, or biomarker
distributions may show multimodality.</p>
<p>We can model SNP effects on <strong>component means</strong> and
<strong>class membership</strong> in a two-Normal mixture.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">p1</span> <span class="op">&lt;-</span> <span class="fl">0.5</span></span>
<span><span class="va">mu1</span> <span class="op">&lt;-</span> <span class="fl">1.2</span>; <span class="va">sd1</span> <span class="op">&lt;-</span> <span class="fl">0.5</span></span>
<span><span class="va">mu2</span> <span class="op">&lt;-</span> <span class="fl">3.0</span>; <span class="va">sd2</span> <span class="op">&lt;-</span> <span class="fl">0.8</span></span>
<span></span>
<span><span class="co"># Simulate phenotype</span></span>
<span><span class="va">z</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/Binomial.html" class="external-link">rbinom</a></span><span class="op">(</span><span class="va">N</span>, <span class="fl">1</span>, <span class="va">p1</span><span class="op">)</span></span>
<span><span class="va">Y</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html" class="external-link">ifelse</a></span><span class="op">(</span><span class="va">z</span> <span class="op">==</span> <span class="fl">1</span>, <span class="fu"><a href="https://rdrr.io/r/stats/Normal.html" class="external-link">rnorm</a></span><span class="op">(</span><span class="va">N</span>, <span class="va">mu1</span>, <span class="va">sd1</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/stats/Normal.html" class="external-link">rnorm</a></span><span class="op">(</span><span class="va">N</span>, <span class="va">mu2</span>, <span class="va">sd2</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">taus</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span><span class="fl">0.1</span>, <span class="fl">0.9</span>, <span class="fl">0.05</span><span class="op">)</span></span>
<span><span class="va">stage1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/quantile_gwas.html">quantile_gwas</a></span><span class="op">(</span><span class="va">Y</span>, <span class="va">G</span>, taus <span class="op">=</span> <span class="va">taus</span><span class="op">)</span></span>
<span><span class="co">#&gt; Building RIF matrix...</span></span>
<span><span class="co">#&gt; Computing per-SNP tau-slopes...</span></span>
<span></span>
<span><span class="va">fit_mix</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/param_gwas.html">param_gwas</a></span><span class="op">(</span></span>
<span>  <span class="va">stage1</span>,</span>
<span>  transform <span class="op">=</span> <span class="st">"two_normal"</span>,</span>
<span>  transform_args <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>    p1 <span class="op">=</span> <span class="va">p1</span>, mu1 <span class="op">=</span> <span class="va">mu1</span>, sd1 <span class="op">=</span> <span class="va">sd1</span>,</span>
<span>    mu2 <span class="op">=</span> <span class="va">mu2</span>, sd2 <span class="op">=</span> <span class="va">sd2</span>,</span>
<span>    include_membership <span class="op">=</span> <span class="cn">TRUE</span></span>
<span>  <span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="va">fit_mix</span><span class="op">$</span><span class="va">params</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt;            gamma        beta_1        beta_2</span></span>
<span><span class="co">#&gt; SNP1 -0.05727378 -1.903029e-02 -0.0233213774</span></span>
<span><span class="co">#&gt; SNP2  0.07954904 -7.991301e-05  0.0076335131</span></span>
<span><span class="co">#&gt; SNP3 -0.06057390 -2.434824e-02  0.0003425944</span></span>
<span><span class="co">#&gt; SNP4  0.02074416 -7.745207e-04 -0.0188630721</span></span>
<span><span class="co">#&gt; SNP5 -0.12819602 -1.526194e-02 -0.0227835128</span></span>
<span><span class="co">#&gt; SNP6  0.11976032  8.322496e-02 -0.0342396586</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="key-functions-and-workflow">Key functions and workflow<a class="anchor" aria-label="anchor" href="#key-functions-and-workflow"></a>
</h2>
<p>The workflow in <code>fungwas</code> always follows <strong>two
stages</strong>:</p>
<ol style="list-style-type: decimal">
<li>
<strong>Quantile GWAS</strong>: estimate SNP effects on <em>quantile
slopes</em> across the phenotype distribution.
<ul>
<li>Function: <code><a href="../reference/quantile_gwas.html">quantile_gwas()</a></code><br>
</li>
<li>Inputs:
<ul>
<li>
<code>Y</code>: vector of phenotypes (length N).<br>
</li>
<li>
<code>G</code>: genotype matrix, N × P (rows = individuals, cols =
SNPs).<br>
</li>
<li>
<code>C</code>: optional covariates (N × K, e.g. age, sex,
PCs).<br>
</li>
<li>
<code>taus</code>: grid of quantile levels
(e.g. <code>seq(0.1, 0.9, 0.05)</code>).<br>
</li>
</ul>
</li>
<li>Output:
<ul>
<li>A list containing RIF slopes per SNP × τ, their SEs, and baseline
quantiles.<br>
</li>
</ul>
</li>
<li>Think of this as a <strong>quantile-level GWAS</strong>.</li>
</ul>
</li>
<li>
<strong>Parameter GWAS</strong>: map quantile slopes into parameter
effects using a weight matrix <code>W</code>.
<ul>
<li>Function: <code><a href="../reference/param_gwas.html">param_gwas()</a></code><br>
</li>
<li>Inputs:
<ul>
<li>The output of <code><a href="../reference/quantile_gwas.html">quantile_gwas()</a></code>.<br>
</li>
<li>A mapping (<code>W</code>) that tells the software how to combine
τ-slopes into parameter effects.<br>
</li>
<li>Either supply a custom <code>W</code> or use a pre-built weight
constructor.<br>
</li>
</ul>
</li>
<li>Output:
<ul>
<li>SNP effects on parameters (means, variances, mixture membership,
etc).<br>
</li>
</ul>
</li>
<li>This is the <strong>interpretation stage</strong>, translating
distributional shifts into biologically meaningful parameters.</li>
</ul>
</li>
</ol>
<div class="section level3">
<h3 id="weight-builders">Weight builders<a class="anchor" aria-label="anchor" href="#weight-builders"></a>
</h3>
<p>Weight matrices (<code>W</code>) define how quantile slopes
correspond to parameter changes.<br><code>fungwas</code> provides several ready-made constructors:</p>
<ul>
<li>
<strong>Variance GWAS (vQTLs)</strong>
<ul>
<li>
<code><a href="../reference/make_weight_vqtl.html">make_weight_vqtl()</a></code> — use when you want SNP effects on
the <em>mean</em> and <em>variance</em> of a Normal trait.<br>
</li>
<li>Example: height variability, BMI dispersion.</li>
</ul>
</li>
<li>
<strong>Mixture GWAS</strong>
<ul>
<li>
<code><a href="../reference/make_weights_normal_mixture.html">make_weights_normal_mixture()</a></code> — use for a two-component
Normal mixture with SNP effects on:
<ul>
<li>Component means.<br>
</li>
<li>(Optionally) class membership probability.<br>
</li>
</ul>
</li>
<li>Example: SNPs shifting balance between subtypes of
cases/controls.</li>
</ul>
</li>
<li>
<strong>Mixture vQTL GWAS</strong>
<ul>
<li>
<code>make_weights_mixture_vqtl()</code> — extended version where
SNPs can also affect the <strong>component variances</strong> (in
addition to means and membership).<br>
</li>
<li>Example: genetic effects on both subtype prevalence and
within-subtype variability.</li>
</ul>
</li>
<li>
<strong>Generic system</strong>
<ul>
<li>
<code><a href="../reference/make_weights_generic.html">make_weights_generic()</a></code> — the most flexible constructor.
<ul>
<li>You provide the distribution’s CDF, PDF, and derivatives wrt
parameters.<br>
</li>
<li>Or use finite-difference helpers
(<code><a href="../reference/make_fd_grad.html">make_fd_grad()</a></code>).<br>
</li>
</ul>
</li>
<li>Example: log-normal GWAS on <code>meanlog</code> and
<code>sdlog</code>.<br>
</li>
<li>Use this when your phenotype is better described by a non-standard
distribution.</li>
</ul>
</li>
</ul>
</div>
<div class="section level3">
<h3 id="putting-it-together-a-typical-workflow">Putting it together: a typical workflow<a class="anchor" aria-label="anchor" href="#putting-it-together-a-typical-workflow"></a>
</h3>
<ol style="list-style-type: decimal">
<li>
<p><strong>Prepare inputs</strong>:</p>
<ul>
<li>Phenotype vector <code>Y</code>.<br>
</li>
<li>Genotype dosage matrix <code>G</code> (SNPs already QC’d).<br>
</li>
<li>Optional covariate matrix <code>C</code> (e.g. sex, age, ancestry
PCs).</li>
</ul>
</li>
<li>
<p><strong>Run quantile GWAS</strong>:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">taus</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span><span class="fl">0.1</span>, <span class="fl">0.9</span>, <span class="fl">0.05</span><span class="op">)</span></span>
<span><span class="va">stage1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/quantile_gwas.html">quantile_gwas</a></span><span class="op">(</span><span class="va">Y</span>, <span class="va">G</span>, taus <span class="op">=</span> <span class="va">taus</span>, C <span class="op">=</span> <span class="va">covariates</span><span class="op">)</span></span></code></pre></div>
</li>
<li>
<p><strong>Choose a weight system</strong>:</p>
<ul>
<li>
<p>If mean &amp; variance:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">W</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/make_weight_vqtl.html">make_weight_vqtl</a></span><span class="op">(</span><span class="va">taus</span>, <span class="va">stage1</span><span class="op">$</span><span class="va">q_tau</span>, mu <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/mean.html" class="external-link">mean</a></span><span class="op">(</span><span class="va">Y</span><span class="op">)</span>, sd <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/sd.html" class="external-link">sd</a></span><span class="op">(</span><span class="va">Y</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
</li>
<li>
<p>If two-component mixture:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">W</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/make_weights_normal_mixture.html">make_weights_normal_mixture</a></span><span class="op">(</span><span class="va">taus</span>, <span class="va">stage1</span><span class="op">$</span><span class="va">q_tau</span>,</span>
<span>                                 p1 <span class="op">=</span> <span class="fl">0.5</span>, mu1 <span class="op">=</span> <span class="fl">1</span>, sd1 <span class="op">=</span> <span class="fl">1</span>,</span>
<span>                                 mu2 <span class="op">=</span> <span class="fl">3</span>, sd2 <span class="op">=</span> <span class="fl">1.5</span>,</span>
<span>                                 include_membership <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
</li>
<li>
<p>If custom:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">W</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/make_weights_generic.html">make_weights_generic</a></span><span class="op">(</span><span class="va">taus</span>, <span class="va">stage1</span><span class="op">$</span><span class="va">q_tau</span>, <span class="va">dist_cdf</span>, <span class="va">dist_pdf</span>, <span class="va">params</span>, <span class="va">grad_funcs</span><span class="op">)</span></span></code></pre></div>
</li>
</ul>
</li>
<li>
<p><strong>Map to parameters</strong>:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">fit</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/param_gwas.html">param_gwas</a></span><span class="op">(</span><span class="va">stage1</span>, transform <span class="op">=</span> <span class="st">"custom_W"</span>, transform_args <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>W <span class="op">=</span> <span class="va">W</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
</li>
<li>
<p><strong>Inspect results</strong>:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="va">fit</span><span class="op">$</span><span class="va">params</span><span class="op">)</span><span class="op">)</span>   <span class="co"># SNP effects on chosen parameters</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="va">fit</span><span class="op">$</span><span class="va">SE_params</span><span class="op">)</span><span class="op">)</span> <span class="co"># Standard errors</span></span></code></pre></div>
</li>
</ol>
<p>In practice:</p>
<ul>
<li>
<strong>Stage 1 (<code>quantile_gwas</code>) is run only
once</strong> per dataset.</li>
<li>You can then re-use it with different weight systems
(<code>param_gwas</code>) to test multiple hypotheses (means, variances,
mixtures, etc) <em>without rerunning the GWAS</em>.</li>
</ul>
</div>
</div>
<div class="section level2">
<h2 id="why-use-fungwas">Why use fungwas?<a class="anchor" aria-label="anchor" href="#why-use-fungwas"></a>
</h2>
<ul>
<li>Go beyond mean effects: test <strong>variance, mixtures,
heterogeneity</strong>.</li>
<li>Fast: closed-form OLS, no heavy quantile regression.</li>
<li>Flexible: any distributional system can be defined via
<code>make_weights_generic</code>.</li>
<li>Useful for <strong>vQTL discovery</strong>, <strong>subtype
genetics</strong>, and <strong>causal inference</strong> when mean
effects don’t tell the full story.</li>
</ul>
</div>
<div class="section level2">
<h2 id="further-reading">Further reading<a class="anchor" aria-label="anchor" href="#further-reading"></a>
</h2>
<ul>
<li>Koenker &amp; Bassett (1978), Quantile Regression.</li>
<li>Firpo, Fortin, &amp; Lemieux (2009), Influence Function
Regression.</li>
<li>Recent applications of vQTL and mixture genetics in GWAS.</li>
</ul>
<hr>
</div>
<div class="section level2">
<h2 id="summary">Summary<a class="anchor" aria-label="anchor" href="#summary"></a>
</h2>
<p><code>fungwas</code> allows you to re-think GWAS: instead of only
asking <em>“does this SNP affect the mean?”</em>, you can ask <em>“does
this SNP affect the distribution?”</em> — variance, mixtures, or other
user-defined systems.</p>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Michel Nivard, Fergus Hamilton.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer>
</div>





  </body>
</html>
