<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>GWAS file handling in R • fungwas</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="96x96" href="../favicon-96x96.png">
<link rel="icon" type="”image/svg+xml”" href="../favicon.svg">
<link rel="apple-touch-icon" sizes="180x180" href="../apple-touch-icon.png">
<link rel="icon" sizes="any" href="../favicon.ico">
<link rel="manifest" href="../site.webmanifest">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="GWAS file handling in R">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-primary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">fungwas</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.0.1</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/fungwas-overview.html">Functional GWAS via Quantile Regression and RIF</a></li>
    <li><a class="dropdown-item" href="../articles/GWAS-in-R.html">GWAS file handling in R</a></li>
    <li><a class="dropdown-item" href="../articles/Q-stat.html">Model Fit Diagnostics with the Q Statistic</a></li>
    <li><a class="dropdown-item" href="../articles/weight-building.html">Your custom functional GWAS with `make_weights_generic`</a></li>
  </ul>
</li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>GWAS file handling in R</h1>
            
      

      <div class="d-none name"><code>GWAS-in-R.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="overview">Overview<a class="anchor" aria-label="anchor" href="#overview"></a>
</h2>
<p>As R isnt very well know for GWAS (most GWAS tools are standalone
commandline apps) we have prepared this vignette. In this vignette we go
trough running a <strong>Quantile GWAS</strong> directly on genotypes
from a PLINK <code>.fam/.bim/.bed</code> dataset, using
<code>bigsnpr</code> to load the genotypes and <code>fungwas</code> for
the quantile regression workflow.</p>
</div>
<div class="section level2">
<h2 id="example-2-gwas-from-plink-files">Example 2: GWAS from PLINK files<a class="anchor" aria-label="anchor" href="#example-2-gwas-from-plink-files"></a>
</h2>
<p>The handling of <strong>PLINK genotype files</strong> in R is done
with the highly optimized tool <code>bigsnpr</code>. For expedience, we
only analyze the first 1000 SNPs. In practice you can chunk the genome
into manageable subsets and itterte over them.</p>
<p>To try this tutorial yourselve you can download plinkfiles to testun
<a href="https://www.dropbox.com/scl/fi/41ep8xbdccp9xw5epim19/1KG.EAS.auto.snp.norm.nodup.split.rare002.common015.missing.zip?rlkey=tklapxwypeg79b1sx03o6ycs7&amp;dl=1" class="external-link">here</a>
its about 200 Mb of files.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># ============================================================</span></span>
<span><span class="co"># Load PLINK bed/bim/fam with bigsnpr</span></span>
<span><span class="co"># Simulate phenotype</span></span>
<span><span class="co"># Run Quantile GWAS with fungwas</span></span>
<span><span class="co"># ============================================================</span></span>
<span></span>
<span><span class="va">plink_prefix</span> <span class="op">&lt;-</span> <span class="st">"/path/to/your-plink/1KG.EAS.auto.snp.norm.nodup.split.rare002.common015.missing/1KG.EAS.auto.snp.norm.nodup.split.rare002.common015.missing"</span>   <span class="co"># without extension</span></span>
<span></span>
<span><span class="co"># Convert to .rds format, runonly  once</span></span>
<span><span class="fu">snp_readBed</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">plink_prefix</span>, <span class="st">".bed"</span><span class="op">)</span><span class="op">)</span>   </span>
<span><span class="va">obj.bigSNP</span> <span class="op">&lt;-</span> <span class="fu">snp_attach</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">plink_prefix</span>, <span class="st">".rds"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">G</span>   <span class="op">&lt;-</span> <span class="va">obj.bigSNP</span><span class="op">$</span><span class="va">genotypes</span>     <span class="co"># Filebacked Big Matrix</span></span>
<span><span class="va">G</span>   <span class="op">&lt;-</span> <span class="fu">snp_fastImputeSimple</span><span class="op">(</span><span class="va">G</span><span class="op">)</span>  <span class="co"># quick mean imputation, for the benefit of simulation</span></span>
<span><span class="va">CHR</span> <span class="op">&lt;-</span> <span class="va">obj.bigSNP</span><span class="op">$</span><span class="va">map</span><span class="op">$</span><span class="va">chromosome</span></span>
<span><span class="va">POS</span> <span class="op">&lt;-</span> <span class="va">obj.bigSNP</span><span class="op">$</span><span class="va">map</span><span class="op">$</span><span class="va">physical.pos</span></span>
<span><span class="va">SNP</span> <span class="op">&lt;-</span> <span class="va">obj.bigSNP</span><span class="op">$</span><span class="va">map</span><span class="op">$</span><span class="va">marker.ID</span></span>
<span><span class="va">N</span>   <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">G</span><span class="op">)</span></span>
<span><span class="va">P</span>   <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">G</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/cat.html" class="external-link">cat</a></span><span class="op">(</span><span class="st">"Loaded"</span>, <span class="va">N</span>, <span class="st">"individuals and"</span>, <span class="va">P</span>, <span class="st">"SNPs\\n"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Simulate phenotype</span></span>
<span><span class="va">target_snps</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="va">P</span>, <span class="fl">5</span><span class="op">)</span></span>
<span><span class="va">beta</span>        <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/Normal.html" class="external-link">rnorm</a></span><span class="op">(</span><span class="fl">5</span>, <span class="fl">0</span>, <span class="fl">0.25</span><span class="op">)</span></span>
<span><span class="va">genetic</span>     <span class="op">&lt;-</span> <span class="fu">big_prodVec</span><span class="op">(</span><span class="va">G</span>, <span class="va">beta</span>, ind.col <span class="op">=</span> <span class="va">target_snps</span><span class="op">)</span></span>
<span><span class="va">y</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/numeric.html" class="external-link">as.numeric</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/scale.html" class="external-link">scale</a></span><span class="op">(</span><span class="va">genetic</span> <span class="op">+</span> <span class="fu"><a href="https://rdrr.io/r/stats/Normal.html" class="external-link">rnorm</a></span><span class="op">(</span><span class="va">N</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Stage 1: Quantile GWAS</span></span>
<span><span class="va">taus</span>   <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span><span class="fl">0.1</span>, <span class="fl">0.9</span>, <span class="fl">0.1</span><span class="op">)</span></span>
<span><span class="va">stage1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/quantile_gwas.html">quantile_gwas</a></span><span class="op">(</span></span>
<span>  Y <span class="op">=</span> <span class="va">y</span>,</span>
<span>  G <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="va">G</span><span class="op">[</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">1000</span><span class="op">]</span><span class="op">)</span>,   <span class="co"># only first 1000 SNPs here</span></span>
<span>  taus <span class="op">=</span> <span class="va">taus</span>,</span>
<span>  benchmark <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Stage 2: Parametric mapping (vQTL)</span></span>
<span><span class="va">W_vqtl</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/make_weights_vqtl.html">make_weights_vqtl</a></span><span class="op">(</span><span class="va">taus</span>, <span class="va">stage1</span><span class="op">$</span><span class="va">q_tau</span>, mu <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/mean.html" class="external-link">mean</a></span><span class="op">(</span><span class="va">y</span><span class="op">)</span>, sd <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/sd.html" class="external-link">sd</a></span><span class="op">(</span><span class="va">y</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">fit</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/param_gwas.html">param_gwas</a></span><span class="op">(</span></span>
<span>  <span class="va">stage1</span>,</span>
<span>  transform <span class="op">=</span> <span class="st">"custom_W"</span>,</span>
<span>  transform_args <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>W <span class="op">=</span> <span class="va">W_vqtl</span><span class="op">)</span>,</span>
<span>  se_mode <span class="op">=</span> <span class="st">"dwls"</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Collect results</span></span>
<span><span class="va">res</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span></span>
<span>  SNP <span class="op">=</span> <span class="va">SNP</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">1000</span><span class="op">]</span>,</span>
<span>  CHR <span class="op">=</span> <span class="va">CHR</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">1000</span><span class="op">]</span>,</span>
<span>  POS <span class="op">=</span> <span class="va">POS</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">1000</span><span class="op">]</span>,</span>
<span>  beta_mu <span class="op">=</span> <span class="va">fit</span><span class="op">$</span><span class="va">params</span><span class="op">[</span><span class="st">"beta_mu"</span>, <span class="op">]</span>,</span>
<span>  beta_sd <span class="op">=</span> <span class="va">fit</span><span class="op">$</span><span class="va">params</span><span class="op">[</span><span class="st">"beta_sigma2"</span>, <span class="op">]</span>,</span>
<span>  se_mu   <span class="op">=</span> <span class="va">fit</span><span class="op">$</span><span class="va">SE_params</span><span class="op">[</span><span class="st">"beta_mu"</span>, <span class="op">]</span>,</span>
<span>  se_sd   <span class="op">=</span> <span class="va">fit</span><span class="op">$</span><span class="va">SE_params</span><span class="op">[</span><span class="st">"beta_sigma2"</span>, <span class="op">]</span>,</span>
<span>  Q       <span class="op">=</span> <span class="va">fit</span><span class="op">$</span><span class="va">Q</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">res</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Simple Manhattan-style plot for mean effect</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/par.html" class="external-link">par</a></span><span class="op">(</span>mar <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">4</span>,<span class="fl">4</span>,<span class="fl">2</span>,<span class="fl">1</span><span class="op">)</span>, bg <span class="op">=</span> <span class="st">"white"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span></span>
<span>  <span class="va">res</span><span class="op">$</span><span class="va">POS</span>,</span>
<span>  <span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log10</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/Chisquare.html" class="external-link">pchisq</a></span><span class="op">(</span><span class="op">(</span><span class="va">res</span><span class="op">$</span><span class="va">beta_mu</span><span class="op">/</span><span class="va">res</span><span class="op">$</span><span class="va">se_mu</span><span class="op">)</span><span class="op">^</span><span class="fl">2</span>, df<span class="op">=</span><span class="fl">1</span>, lower.tail <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span><span class="op">)</span>,</span>
<span>  col <span class="op">=</span> <span class="va">res</span><span class="op">$</span><span class="va">CHR</span>, pch <span class="op">=</span> <span class="fl">20</span>, cex <span class="op">=</span> <span class="fl">0.6</span>,</span>
<span>  xlab <span class="op">=</span> <span class="st">"Genomic position"</span>, ylab <span class="op">=</span> <span class="st">"-log10(p)"</span>,</span>
<span>  main <span class="op">=</span> <span class="st">"Quantile GWAS (mu effect)"</span></span>
<span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="chunking-the-gwas">Chunking the GWAS<a class="anchor" aria-label="anchor" href="#chunking-the-gwas"></a>
</h2>
<p>In this example we only analyzed the first 1000 SNPs for speed. In
practice, you can loop over chunks of SNPs and process them in parallel
using <code>future.apply</code> or <code>big_apply</code> from
<strong>bigsnpr</strong>. The workflow is the same:</p>
<ol style="list-style-type: decimal">
<li>Select a block of SNPs from the <code>G</code> matrix.</li>
<li>Run <code><a href="../reference/quantile_gwas.html">quantile_gwas()</a></code> on that block.</li>
<li>Map results with <code><a href="../reference/param_gwas.html">param_gwas()</a></code>.</li>
<li>Save results per chunk, then combine.</li>
</ol>
<p>This strategy readily scales the quantile GWAS to genome-wide
data.</p>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Michel Nivard, Fergus Hamilton.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer>
</div>





  </body>
</html>
